File: app/src/main/java/me/pwcong/jpstart/component/fragment/JPStartFragment.java
package me.pwcong.jpstart.component.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

import java.util.List;

import butterknife.BindView;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.adapter.JPStartRecyclerAdapter;
import me.pwcong.jpstart.conf.Constants;
import me.pwcong.jpstart.manager.SoundPoolManager;
import me.pwcong.jpstart.mvp.bean.JPItem;
import me.pwcong.jpstart.mvp.presenter.BasePresenter;
import me.pwcong.jpstart.mvp.presenter.JPStartFragmentPresenterImpl;
import me.pwcong.jpstart.mvp.view.BaseView;
import me.pwcong.jpstart.utils.ResourceUtils;
import me.pwcong.jpstart.widget.ImageDialog;

/**
 * Created by Pwcong on 2016/9/27.
 */

public class JPStartFragment extends BaseFragment implements BaseView.JPStartFragmentView{

    @BindView(R.id.recycler_view)
    RecyclerView mRecyclerView;

    BasePresenter.JPStartFragmentPresenter presenter;

    JPStartRecyclerAdapter adapter;

    int category_yin = 0;

    @Override
    protected int getViewId() {
        return R.layout.fragment_jpstart;
    }

    public static JPStartFragment newInstance(int type){

        Bundle argument=new Bundle();
        argument.putInt(Constants.CATEGORY_YIN,type);

        JPStartFragment fragment=new JPStartFragment();
        fragment.setArguments(argument);

        return fragment;

    }

    @Override
    protected void initVariable(@Nullable Bundle savedInstanceState) {

        category_yin = getArguments().getInt(Constants.CATEGORY_YIN);

        presenter=new JPStartFragmentPresenterImpl(this);

    }


    @Override
    protected void doAction() {
        presenter.initJPStartFragment(category_yin);
    }

    @Override
    public void setData(List<JPItem> data) {

        adapter=new JPStartRecyclerAdapter(data);
        adapter.setOnItemClickListener(new JPStartRecyclerAdapter.OnItemClickListener() {
            @Override
            public void onClick(JPItem item) {
                SoundPoolManager.getInstance().play(item.getRome());
            }
        });

        adapter.setOnItemLongClickListener(new JPStartRecyclerAdapter.OnItemLongClickListener() {
            @Override
            public void onLongClick(JPItem item) {
                new ImageDialog.Builder(getContext())
                        .setResId(R.raw.write_a)
                        .override((int)ResourceUtils.getDimension(getContext(),R.dimen.dialog_width),
                                (int)ResourceUtils.getDimension(getContext(),R.dimen.dialog_height))
                        .create()
                        .show();
            }
        });

        mRecyclerView.setAdapter(adapter);

    }

    @Override
    public void setRecyclerView(int type) {

        switch (type){
            case Constants.CATEGORY_QINGYIN:
                RecyclerView.LayoutManager layoutManager1=new GridLayoutManager(getContext(),Constants.COLUMN_QINGYIN);
                mRecyclerView.setLayoutManager(layoutManager1);
            case Constants.CATEGORY_ZHUOYIN:
                RecyclerView.LayoutManager layoutManager2=new GridLayoutManager(getContext(),Constants.COLUMN_ZHUOYIN);
                mRecyclerView.setLayoutManager(layoutManager2);
                break;
            case Constants.CATEGORY_AOYIN:
                RecyclerView.LayoutManager layoutManager3=new GridLayoutManager(getContext(),Constants.COLUMN_AOYIN);
                mRecyclerView.setLayoutManager(layoutManager3);
                break;
            default:break;
        }

    }
}

File: app/src/main/java/com/github/pwcong/jpstart/ui/fragment/JPStartFragment.kt
package com.github.pwcong.jpstart.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.pwcong.jpstart.App
import com.github.pwcong.jpstart.R
import com.github.pwcong.jpstart.adapter.JPStartRecyclerAdapter
import com.github.pwcong.jpstart.constant.Constant
import com.github.pwcong.jpstart.databinding.FragmentJpstartBinding
import com.github.pwcong.jpstart.manager.GifManager
import com.github.pwcong.jpstart.manager.SoundPoolManager
import com.github.pwcong.jpstart.mvp.bean.JPItem
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter.JPStartFragmentPresenter
import com.github.pwcong.jpstart.mvp.presenter.impl.JPStartFragmentPresenterImpl
import com.github.pwcong.jpstart.mvp.view.BaseView.JPStartFragmentView
import com.github.pwcong.jpstart.ui.component.dialog.ImageDialog
import com.github.pwcong.jpstart.utils.ResourceUtils

class JPStartFragment : BaseFragment<FragmentJpstartBinding>(), JPStartFragmentView {

    private lateinit var mRecyclerView: RecyclerView

    private lateinit var presenter: JPStartFragmentPresenter

    private var categoryYin = 0

    override fun initViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?
    ): FragmentJpstartBinding {
        return FragmentJpstartBinding.inflate(inflater, container, false)
    }

    override fun initVariable(savedInstanceState: Bundle?) {
        mRecyclerView = getViewBinding().recyclerView

        categoryYin = requireArguments().getInt(Constant.CATEGORY_YIN)

        presenter = JPStartFragmentPresenterImpl(this)
    }

    override fun doAction() {
        presenter.initJPStartFragment(categoryYin)
    }

    override fun setData(data: List<JPItem>) {
        val adapter = JPStartRecyclerAdapter(data)

        adapter.setOnItemClickListener(object : JPStartRecyclerAdapter.OnItemClickListener {
            override fun onClick(item: JPItem) {
                SoundPoolManager.getInstance().play(item.rome)
            }
        })

        adapter.setOnItemLongClickListener(object :
            JPStartRecyclerAdapter.OnItemLongClickListener {
            override fun onLongClick(item: JPItem) {
                val gif = GifManager.getInstance().getJPGif(item.rome)
                if (gif != null) {
                    if (App.TYPE_MING == Constant.TYPE_HIRAGANA) {
                        context?.let {
                            ImageDialog.Builder(it)
                                .setResId(gif.hiragana)
                                .override(
                                    ResourceUtils.getDimension(it, R.dimen.dialog_width).toInt(),
                                    ResourceUtils.getDimension(it, R.dimen.dialog_height).toInt()
                                )
                                .create()
                                .show()
                        }
                    } else {
                        context?.let {
                            ImageDialog.Builder(it)
                                .setResId(gif.katakana)
                                .override(
                                    ResourceUtils.getDimension(it, R.dimen.dialog_width).toInt(),
                                    ResourceUtils.getDimension(it, R.dimen.dialog_height).toInt()
                                )
                                .create()
                                .show()
                        }
                    }
                }
            }
        })

        mRecyclerView.adapter = adapter
    }

    override fun setRecyclerView(type: Int) {
        when (type) {
            Constant.CATEGORY_QINGYIN -> {
                val layoutManager: RecyclerView.LayoutManager = GridLayoutManager(
                    context,
                    Constant.COLUMN_QINGYIN
                )
                mRecyclerView.layoutManager = layoutManager
            }

            Constant.CATEGORY_ZHUOYIN -> {
                val layoutManager: RecyclerView.LayoutManager = GridLayoutManager(
                    context,
                    Constant.COLUMN_ZHUOYIN
                )
                mRecyclerView.layoutManager = layoutManager
            }

            Constant.CATEGORY_AOYIN -> {
                val layoutManager: RecyclerView.LayoutManager =
                    GridLayoutManager(context, Constant.COLUMN_AOYIN)
                mRecyclerView.layoutManager = layoutManager
            }

            else -> {}
        }
    }

    companion object {
        fun newInstance(type: Int): JPStartFragment {
            val argument = Bundle()
            argument.putInt(Constant.CATEGORY_YIN, type)

            val fragment = JPStartFragment()
            fragment.arguments = argument

            return fragment
        }
    }
}
